src/sagas/walletConnect.js

```diff
@@ -0,0 +1,494 @@
+      response,
+    }));
+  } catch (error) {
+    console.log('Captured error on signMessage: ', error);
+    yield put(onExceptionCaptured(error));
+  }
+}
+
+/**
+ * Listens for the wallet reset action, dispatched from the wallet sagas so we
+ * can clear all current sessions.
+ */
+export function* onWalletReset() {
+  const walletConnect = yield select((state) => state.walletConnect);
+  if (!walletConnect || !walletConnect.client) {
+    // Do nothing, wallet connect might not have been initialized yet
+    return;
+  }
+
+  yield call(clearSessions);
+}
+
+/**
+ * This saga will be called (dispatched from the event listener) when a session
+ * proposal RPC is sent from a dApp. This happens after the client scans a wallet
+ * connect URI
+ */
+export function* onSessionProposal(action) {
+  const { id, params } = action.payload;
+  const { web3wallet } = yield select((state) => state.walletConnect.client);
+
+  const wallet = yield select((state) => state.wallet);
+  const firstAddress = yield call(() => wallet.getAddressAtIndex(0));
+
+  const data = {
+    icon: get(params, 'proposer.metadata.icons[0]', null),
+    proposer: get(params, 'proposer.metadata.name', ''),
+    url: get(params, 'proposer.metadata.url', ''),
+    description: get(params, 'proposer.metadata.description', ''),
+    requiredNamespaces: get(params, 'requiredNamespaces', []),
+  };
+
+  const onAcceptAction = { type: 'WALLET_CONNECT_ACCEPT' };
+  const onRejectAction = { type: 'WALLET_CONNECT_REJECT' };
+
+  yield put(setWalletConnectModal({
+    show: true,
+    type: WalletConnectModalTypes.CONNECT,
+    data,
+    onAcceptAction,
+    onRejectAction,
+  }));
+
+  const { reject } = yield race({
+    accept: take(onAcceptAction.type),
+    reject: take(onRejectAction.type),
+  });
+
+  if (reject) {
+    yield call(() => web3wallet.rejectSession({
+      id: params.id,
+      reason: {
+        code: ERROR_CODES.USER_REJECTED,
+        message: 'User rejected the session',
+      },
+    }));
+  }
+
+  try {
+    yield call(() => web3wallet.approveSession({
+      id,
+      relayProtocol: params.relays[0].protocol,
+      namespaces: {
+        hathor: {
+          accounts: [`hathor:${NETWORK}:${firstAddress}`],
+          chains: [`hathor:${NETWORK}`],
+          events: AVAILABLE_EVENTS,
+          methods: values(AVAILABLE_METHODS),
+        },
+      },
+    }));
+
+    yield call(refreshActiveSessions);
+  } catch (error) {
+    console.error('Error on sessionProposal: ', error);
+    yield put(onExceptionCaptured(error));
+  }
+}
+
+/**
+ * This saga is fired when a URI is inputted either manually or by scanning
+ * a QR Code
+ */
+export function* onUriInputted(action) {
+  const { web3wallet, core } = yield select((state) => state.walletConnect.client);
+
+  if (!web3wallet) {
+    throw new Error('Wallet connect instance is new and QRCode was read');
+  }
+
+  const { payload } = action;
+
+  try {
+    yield call(core.pairing.pair, { uri: payload });
```

Comment:
What's this call?

Reply:
This is a redux-saga call effect, it is blocking meaning that it will wait the `core.pairing.pair` promise finish before continuing the generator

Reply:
Ah yeah but what's the method `core.pairing.pair` doing?

Reply:
It's called when we read a QR code or input an URI manually, it will initialize a session with the information from the QR Code and ask for user approval, I added a flow explanation in e634d2e04631972c316c060df4d659a375b37d22